Unify Quantization APIs for add, pool and relu

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26335 Unify Quantization APIs for add, pool and relu**
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK

Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504331](https://our.internmc.facebook.com/intern/diff/D17504331)

supriyar committed:
Unify Quantization APIs for add, pool and relu
Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -67,9 +70,83 @@ Tensor _add_scalar_out(Tensor& out, const Tensor& self, Scalar other) {
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK Add operator");
+  TORCH_INTERNAL_ASSERT(qnnpack_operator != nullptr);
```
nit: error message? also if createStatus == success, is this assert always true?

supriyar replied:
Yeah, this is just an additional check. Can be removed.

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qadd.cpp:
```diff
@@ -67,9 +70,83 @@ Tensor _add_scalar_out(Tensor& out, const Tensor& self, Scalar other) {
+  const pytorch_qnnp_status runStatus =
+      pytorch_qnnp_run_operator(qnnpack_operator, threadpool);
+
+  TORCH_INTERNAL_ASSERT(
```
I think this one might not be INTERNAL ASSERT since user might have invalid input etc.?

Also looks like L131-L137 are general to all qnnpack ops? should we refactor this to separate function?

supriyar replied:
We check for inputs separately at the start. This is an internal assert from something going wrong inside qnnpack.
I don't think refactoring line 131-137 is necessary at this point, as eventually we will change all ops API for qnnpack to make it more functional

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -244,12 +247,141 @@ Tensor quantized_max_pool2d(
+         "qnnpack_maxpool(): Expected padding to be 2-dimensional: got ",
+         padding.size());
+
+     Tensor input_contig = input.contiguous();
```
Do you need a check for input layout here? How do you know that the input is nhwc contiguous?

supriyar replied:
I'll make the change similar to https://github.com/pytorch/pytorch/pull/26242 once that lands for all qnnpack ops.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1222,18 +1222,18 @@ class TestQNNPackOps(TestCase):
-
-        X = torch.from_numpy(X)
-        Y = X.clone()
+        with enable_mobile_quantized_engine():
```
Similar comments to conv/linear qnnpack tests:  We should ensure that all backends share the same tests and have consistent numerics. Can you merge this test with the existing relu test?

supriyar replied:
Discussed offline, will merge in a separate PR.

raghuramank100 reviewed (changes requested):
I have comments mainly on test coverage and on unifying tests between FBGEMM and QNNPACK.

supriyar commented on test/test_quantized.py:
```diff
@@ -1541,40 +1541,41 @@ def test_qconv_unpack(self, X, strideH, strideW, padH, padW):
-        B, (scale_b, zero_point_B, torch_type) = A_temp
-        A = torch.from_numpy(A)
-        B = torch.from_numpy(B)
+        with enable_mobile_quantized_engine():
```
Will update the zero point. The input is quint8 since that is the only data type qnnpack currently supports.

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

t-vi commented:
It might well be me, but on raspbian (32bit armv7 hf) and `USE_PYTORCH_QNNPACK=1`, I get None from conv_prepack called like this:
```
r = torch.ones([2,2,2,2], dtype=torch.float)
scale, zero_point = 1, 2
q = torch.quantize_linear(r, scale, zero_point, torch.qint8)
torch.ops.quantized.conv_prepack(q, None, [1,1], [0,0], [1,1], 1)
```
I do get some torch.uint8 tensor when I do the same with a vanilla build on x86 (i.e. with FBGEMM).

supriyar commented:
> It might well be me, but on raspbian (32bit armv7 hf) and `USE_PYTORCH_QNNPACK=1`, I get None from conv_prepack called like this:
> 
> ```
> r = torch.ones([2,2,2,2], dtype=torch.float)
> scale, zero_point = 1, 2
> q = torch.quantize_linear(r, scale, zero_point, torch.qint8)
> torch.ops.quantized.conv_prepack(q, None, [1,1], [0,0], [1,1], 1)
> ```
> 
> I do get some torch.uint8 tensor when I do the same with a vanilla build on x86 (i.e. with FBGEMM).

Did you try setting torch.backends.quantized.engine = torch.qnnpack? You can set the same on server side (x86) as well to confirm qnnpack works.

t-vi commented:
Indeed, this fixes the example case and seems to return something in the quantized model. Thank you!

At some point, `conv_prepack` might raise a more descriptive exception except of just returning None... :slightly_smiling_face: 
Should the engine default to qnnpack when it available but fbgemm isn't?

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qpool.cpp:
```diff
@@ -244,12 +247,142 @@ Tensor quantized_max_pool2d(
+     int64_t inW = input.size(3);
+     // TODO: change it to contiguous(MemoryFormat::ChannelsLast) once a perf
+     // regression of it is fixed.
+     Tensor input_contig = input.permute({0, 2, 3, 1}).contiguous();
```
@dzhulgakov please review this. I changed it to permute input tensor to NHWC similar to your changes for qconv.

dzhulgakov replied:
Yep, it should be fine

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

dzhulgakov reviewed (commented):
Looks good, I'll let Jerry to review in the details.

Would be really nice to unify the tests too and extend the benchmarks in benchmarks/op_benchmarks to go through both engines

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LGTM

supriyar committed:
Update on "Unify Quantization APIs for add, pool and relu"


Summary:
Use the backend engine flag to call QNNPACK for quantized ops.

Test Plan:
python test/test_quantized.py TestQNNPACKOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17504331](https://our.internmc.facebook.com/intern/diff/D17504331)

[ghstack-poisoned]
